### ISSUE TITLE
20171104-20171117:Lec12-13

### ISSUE THREAD

**Issue by stuian** [Neuqmiao/deepNLP#4]
## 此issue为本周提交作业以及提问讨论的区域
提交作业的形式为：

- 姓名：xxx
- 作业：链接
- 感想：可以对本周学习内容进行小总结，或者学习过程中的感受，也可以是学习过程中的问题等等。

**Comment by wtybill1** [comment]
姓名：陈梓涵
内容：LEC12
感想：
1.LEC12介绍了Memory Networks，这种模型解决了RNN、LSTM记忆太小的问题，加了一个m模块，和I、G、O、R四个元件（从文本中提取输入输入到G中，结合记忆，根据问题给出输出，然后再把向量化转变为原本的格式，返回答案）（似乎有点像各种门）
2.在Memory Networks中，作者在基础的模型上加了一层segemnt，如果没理解错的话里面的每个模块都可以更变，这种模型的根本还是RNN。
3.尝试地推导了一下公式，没导出来（应该是符号没弄清楚的原因。。）

**Comment by WALLE499** [comment]
姓名：赵松
https://github.com/WALLE499/CS224D_NLP.git
感想：
1、lec12的内容是 memory network 有点类似于之前学习的LSTM，可以解决长时记忆问题，模型的目标是是根据文本内容回答问题。
2、memory network 包括四部分I、G、O、R，
I：可以进行前期预处理，G：可以存储记忆或者选择遗忘，O：根据新的输入值还有记忆值产生新的输出，R：将输出转换为所需的响应格式。
3、看了PPT还有一些中文博客的笔记，时间比较紧lec12的论文没有详细的读。
4、lec13的内容是关于CNN在NLP中的应用，与RNN不同的是CNN不需要语法分析器。
5、lec13中还介绍了一些提高CNN性能的技巧。

**Comment by Lucas-Fan** [comment]
姓名：樊志英
感想：
1.lec12的介绍了CNN，并讲了其在自然语言翻译中的应用，还介绍了Bag of Vectors、Window	Model和CNNs的比较。
2.这周还学习了NLTK框架，跟着《Python自然语言处理》这本书学，还了解了python包jieba用来中文分词，BeautifulSoup包是一个非常强大的文本处理的包。学会了基本的文本处理和分词操作。（练习的代码在https://github.com/Lucas-Fan/NLTK）

**Comment by stuian** [comment]
姓名：蒋文君
作业：[链接](https://github.com/stuian/NLP/blob/master/lec12/note/Lec12.md)
感想：
memory networks有两部分，inference components和long-term memory components(在别人的基础上新增的:可以被用来读取和写入)
论文中：

- 第二部分介绍了memory networks的整体框架；
- 第三部分展示了一个文本领域的问答任务的确切应用；
- 第四部分讨论了相关的其他论文；
- 第五部分描述了我们的试验；第六部分最后进行总结。

第一部分了解的差不多，第二部分很多细节还不明白。

**Comment by zjp96machineL** [comment]
Lec12memary-networks

第十二讲其实主要是引入了一个memary-network的框架，包含四大组件，并且针对其中的各组件逐一解释说明，同时文章也举了一个简单的应用于问答系统的例子，最后又对各部件进行拓展。

Memary-network产生原因：rnn，lstm，gru等当前的记忆机制都把记忆嵌入一个低维度的空间，丢失很多信息，无法满足大容量记忆的需要。
组件：I：将输入转变为内部的特征表达。
G：利用新的输入更新记忆。
O：利用新的记忆给出输出。
R：将输出转变为期望的回应格式。

一个memNN的实例：
文本问答：输入句子，直接以最原始的结构存到记忆单元。当输入一个问题x时，则利用得分函数从记忆单元中选择得分最高的句子s1，再利用s1，x，选出得分最高的句子s2，然后再利用x，s1，s2选出一个得分最高的词，作为输出（限定了输出为词）。Loss function为一个最大间隔损失函数，还利用了负采样，训练前各个匹配的句子和词已经被标记。
另外，该实例的数据集是手工打造的，利用简单的语法结构，加入了简单的合并陈述。存在问题：指代，复合名词等，与真实数据差异很大！

拓展：
1.输入是词序列怎么办。方法：利用得分函数做分割。
2.hash搜索（没接触过，不清楚具体的实施）
3.考虑到input的顺序对预测的影响（例如一个故事作为输入，其中各个句子的输入顺序就会对预测产生影响），方法：增加维度表示顺序。
4.输入有未出现的词该怎么办，方法：增加维度表示上下文，利用上下文去代替。

Lec13 CNN

本讲主要是讲述了CNN在NLP的应用。
Cnn与rnn的对比：
1.cnn与rnn的步长不同，cnn的步长更为复杂，如1，2等等
2.cnn有多个filter，并具有pooling结构
3.cnn输入长度受限需要加padding，rnn没有
4.cnn每层的filter权值不变的（不同种类的filter不同，但每个filter对于与他对应的所有窗口的都是相同的），rnn全部相同。
Recurrent neural network：循环神经网络，按时间维度展开，无法利用未来的特征，受最后的词向量影响大。
Recursive neural-network：递归神经网络，按空间维度展开，需要语法解析器帮助。
CNN，可以考虑到每个可能的短语，而rnn只考虑符合语法的短语。

首先介绍了一维的离散卷积，相当于在一维的直线上提取特征，另外又介绍了二维卷积。基本的卷积过程就不再赘述了。

利用cnn处理句子的基本流程如下：
1.	输入时级联的词向量，利用filter W去按步长依次遍历，得到多个节点表式。
2.	Max-Polling，取最大。
存在的问题：这样做得到的特征太少了，解决方案：加入多个filters，由于是max-pooling操作，所以不用考虑filter的窗口大小，可以是bigram    4-gram等等。
训练的细节：
1.muli-channel idea开始时，利用两份完全一样的词向量（两个channel）
在反向传播时仅仅时修改其中一份，另一份保持固定
每次都把两个通道生成的节点值相加。
为什么引入多通道呢？
个人观点是在训练时我们只是在训练一个局部的样本，词向量会根据目标分类而变动，任由梯度流入，会使得模型失去它在语义上的泛化行，类似于防止过拟合。
2.Dropout
3.正则化
对输出权值矩阵每一行的2-范数进行数值上的限制。不常用。

**Comment by zjp96machineL** [comment]
张金鹏
总结:
Lec12memary-networks

第十二讲其实主要是引入了一个memary-network的框架，包含四大组件，并且针对其中的各组件逐一解释说明，同时文章也举了一个简单的应用于问答系统的例子，最后又对各部件进行拓展。

Memary-network产生原因：rnn，lstm，gru等当前的记忆机制都把记忆嵌入一个低维度的空间，丢失很多信息，无法满足大容量记忆的需要。
组件：I：将输入转变为内部的特征表达。
G：利用新的输入更新记忆。
O：利用新的记忆给出输出。
R：将输出转变为期望的回应格式。

一个memNN的实例：
文本问答：输入句子，直接以最原始的结构存到记忆单元。当输入一个问题x时，则利用得分函数从记忆单元中选择得分最高的句子s1，再利用s1，x，选出得分最高的句子s2，然后再利用x，s1，s2选出一个得分最高的词，作为输出（限定了输出为词）。Loss function为一个最大间隔损失函数，还利用了负采样，训练前各个匹配的句子和词已经被标记。
另外，该实例的数据集是手工打造的，利用简单的语法结构，加入了简单的合并陈述。存在问题：指代，复合名词等，与真实数据差异很大！

拓展：
1.输入是词序列怎么办。方法：利用得分函数做分割。
2.hash搜索（没接触过，不清楚具体的实施）
3.考虑到input的顺序对预测的影响（例如一个故事作为输入，其中各个句子的输入顺序就会对预测产生影响），方法：增加维度表示顺序。
4.输入有未出现的词该怎么办，方法：增加维度表示上下文，利用上下文去代替。

Lec13 CNN

本讲主要是讲述了CNN在NLP的应用。
Cnn与rnn的对比：
1.cnn与rnn的步长不同，cnn的步长更为复杂，如1，2等等
2.cnn有多个filter，并具有pooling结构
3.cnn输入长度受限需要加padding，rnn没有
4.cnn每层的filter权值不变的（不同种类的filter不同，但每个filter对于与他对应的所有窗口的都是相同的），rnn全部相同。
Recurrent neural network：循环神经网络，按时间维度展开，无法利用未来的特征，受最后的词向量影响大。
Recursive neural-network：递归神经网络，按空间维度展开，需要语法解析器帮助。
CNN，可以考虑到每个可能的短语，而rnn只考虑符合语法的短语。

首先介绍了一维的离散卷积，相当于在一维的直线上提取特征，另外又介绍了二维卷积。基本的卷积过程就不再赘述了。

利用cnn处理句子的基本流程如下：
1.	输入时级联的词向量，利用filter W去按步长依次遍历，得到多个节点表式。
2.	Max-Polling，取最大。
存在的问题：这样做得到的特征太少了，解决方案：加入多个filters，由于是max-pooling操作，所以不用考虑filter的窗口大小，可以是bigram    4-gram等等。
训练的细节：
1.muli-channel idea开始时，利用两份完全一样的词向量（两个channel）
在反向传播时仅仅时修改其中一份，另一份保持固定
每次都把两个通道生成的节点值相加。
为什么引入多通道呢？
个人观点是在训练时我们只是在训练一个局部的样本，词向量会根据目标分类而变动，任由梯度流入，会使得模型失去它在语义上的泛化行，类似于防止过拟合。
2.Dropout
3.正则化
对输出权值矩阵每一行的2-范数进行数值上的限制。不常用。

**Comment by hzc5095** [comment]
郝增超
链接：https://github.com/hzc5095/cs224d
总结：主要是讲了memory network的工作原理，以及使用的一些公式介绍。
大致就是（QA举例子）把一段话变成词向量，按顺序存入数组中，提出问题同样转化为词向量找到与他最相似的记忆，作为答案。
心得：对一些公式的具体运算流程不太熟，只知道怎么用。最近考试有点多，打算找个例子实验一下，还没有时间找。
具体就是上传的PDF文件。

**Comment by Hotown** [comment]
姓名：罗宏涛
作业：http://ihotown.cn/article/28
总结：
lec12简述了Memory Networks模型，在例子中主要用来解决Q&A问题。

之所以会提出这个模型，原因在于，之前模型的memory受到限制，实现一些"low level"的任务，比如对象预测。我理解的是，简单模型的memory实际上是保存在各个权重矩阵中，学习权重的过程实际上就可以看做是memory的作用，就像人类在学习的过程中，memory也占据着很大的比重。

但是要解决Q&A问题，单靠简单的memory可能无法完成，因此才提出Memory Networks。

MemNN有四个组件，I，G，O，R

+ I（input feature map）：用于处理输入，转化成特征表示
+ G（generalization）：通过新的输入来更新现有的memory
+ O（output feature map）：根据新的memory重新计算输出
+ R（response）：将输出转化为文本化的响应

**Comment by Sicanno** [comment]
姓名：刘云珂  
作业：https://github.com/Sicanno/cs224d/tree/lec-13-cnn  
感想：  
1.lec 12主要讲了针对QandA问题，memory networks模型大致看了一下，没有仔细了解。  
2.lec 13讲了cnn，从简单模型到复杂模型，主要对于刚开始生成两个maps的过程不太理解以及multiple feature maps的过程。

**Comment by gitabtion** [comment]
- 姓名：蔡珩
- 作业：
- 感想：

	- 本周工作

		- 完成了NEUQ-DELTA实验室服务器环境的进一步完善工作
		- 学习了神经网络的概念及其评价方法
		- 学习了前馈神经网络及误差反向传播的思想及方法
		- 学习了卷积神经网络模型并深入讨论了反卷积的作用及意义
		- 了解了Memory Networks模型的概念及思想
		
	- 下周计划
		- 学习BP神经网络
		- 利用Tensorflow实现简单的cnn神经网络搭建
		- 利用TensorFlow实现手写数字识别
		- 了解SLING框架，并阅读源码

	- 存在的问题
		- 神经网络在误差逆向传播的过程中，权值等的调整是通过怎样的方式进行的
		- relu和sigmod两个函数的优劣在何种情况下更为明显，或者说哪个函数在哪种环境下更为适用，relu的宽大的值域是否不符合压缩的概念。
		- 就目前神经网络相关的论文而言，很多问题都已有较好的模型去解决，而目前市场上对于利用机器学习模型及算法来解决这些问题的应用并不太多，是否存在部分论文的结果的局限性或者科学理论成果的市场转化不到位等问题。

		


# Lecture12 Memory Networks

## Abstract


- reason with inference components combianed with a long-term memory components
- long-term memory components 
	- can be read and write to
	- using it to prediction
	- effectively as a (dynamic) knowledge base


## Introduction

- 大多数传统的机器学习算法
	- 没有好的方法去优雅地操作long-term memory
	- 无法充分利用现代计算机的强大的计算能力
	- 记忆能力差，且不精确

- Memory Networks

	- m: memory array of object
	- I: (input feature map) 
		- converts the incoming input to the internal feature representation.
	- G: (generalization) 
		- updates old memories given the new input. We call this generalization as there is an opportunity for the network to compress and generalize its memories at this stage for some intended future use.
	- O: (output feature map)
		- produces a new output (in the feature representation space), given the new input and the current memory state.
	- R: (response)
		- converts the output into the response format desired. 

	- 在得到输入之后，memory networks 的训练和测试流程主要分为以下几步：

		1. Convert x to an internal feature representation I(x).
		2. Update memories mi given the new input: m\_i = G(m\_i, I(x), m), ∀i.
		3. Compute output features o given the new input and the memory: o=O(I(x),m).
		4. Finally, decode output features o to give the final response: r = R(o).
	- 当训练集与测试集产生差别时，m会实时存储，但I,G,O,R四个组件不会更改
	- I可以使用标准预处理后输入
	- m_H(x) = I(x)
	- H可以对每个memory评分在，memory满了的情况下替换掉评分最低的memory

**Comment by Lucas-Fan** [comment]
姓名：樊志英
报告：https://github.com/Lucas-Fan/lec12-13
有一些图片在讨论里面不好排版，将报告传到github上了，上面是网址。

**Comment by millllk** [comment]
姓名：吕思奇
学习内容：lec12课程和两篇论文
                 lec13还没有看完
总结：lec12主要介绍了memory networks 是一个有监督学习 用于解决基于给定文字回答问题，主要原理是利用QDMatch来找到可以用于回答问题的supporting facts 比较简单的问题只需要1个supporting fact 复杂的就需要两个及两个以上的了 MemNN主要有4个步骤：I G O R 最核心的算法在O这一步骤 这个网络还可以解决与时间有关的问题

**Comment by Lonny97** [comment]
姓名：王历宁
感想：
lec12的内容是 memory network 类似LSTM，解决长时记忆问题，用tensorflow实现了一遍RNN解决回归问题和分类问题
memory network 包括四部分I、G、O、R，
I（input feature map）：进行前期预处理，输入数据转化成特征表示，G（generalization）：可以存储记忆或者选择遗忘，O（output feature map）：根据新的输入值还有记忆值产生新的输出，R（response）：将输出转化为文本化的响应
lec13介绍了一些CNN在NLP中的应用，用tensorflow手动实现了一遍CNN，对每个卷积核矩阵的初值不是很了解。

今晚请假，英语6级补课
